### ISSUE TITLE
Task slick.backend.DatabaseComponent rejected from java.util.concurrent.ThreadPoolExecutor

### ISSUE THREAD

**Issue by Salmondx** [slick/slick#1183]
I'm using slick 3.0.0 version.
Exception: 
```
java.util.concurrent.RejectedExecutionException: Task slick.backend.DatabaseComponent$DatabaseDef$$anon$2@33f65f2 rejected from java.util.concurrent.ThreadPoolExecutor@1311d5eb[Running, pool size = 20, active threads = 20, queued tasks = 1000, completed tasks = 73]
```
As i think, this problem occurred due to lack of free threads in thread pool but i have no idea how to increase them. In hikariCP for my computer spec i need exactly 20 threads in pool. 
Any recommendations?

**Comment by tonylotts** [comment]
I'm getting this exception thrown in specs.

**Comment by prettynatty** [comment]
I also get this exception in tests on a simple UPDATE query with one WHERE.
As a work-around I rewrote action using plain SQL.

**Comment by prettynatty** [comment]
A work-around I mentioned above didn't work. It fixed one test, but the problem migrated to another one! :poodle:

**Comment by prettynatty** [comment]
I caught this problem when I run `DBIO.fromFuture` inside a transactionally DBIOAction.

**Comment by sunil-123** [comment]
I'm getting same exception when I'm running specs. I'm using `H2` database for specs.

**Comment by alexlee85** [comment]
I caught this problem when I use transaction like db.run(action.transactionally). if I remove the transactionally, it will be fine.

**Comment by maesenka** [comment]
Same here: large number of transactional DBIOActions in quick succession triggered the same exception. Removing the .transactionally makes the problem go away, but I'd like to keep the transactional guarantee.

**Comment by szeiger** [comment]
This is the intended behavior. It happens when too many actions queue up. You can increase the `queueSize` in the database configuration to avoid it (there was a bug fix for resolving the correct config value in 3.0.2). The default is 1000.

However, I have no idea why it would behave differently for transactional vs auto-commit actions. Maybe the transactional versions are simply slower in the database so you reach the limit sooner? Slick prioritizes all follow-on actions (e.g. the second part of an `andThen` or `flatMap`) over new actions so that running action sequences do not get rejected at a later point, but this should not be any different for transactions (`a.transactionally` is just `StartTransaction.andThen(a).cleanUp(eo => if(eo.isEmpty) Commit else Rollback)`)

**Comment by calippo** [comment]
@szeiger I'm having the same issue, and I'd like to change the rejection policy. I tried with `property`: `rejection-policy = "caller-runs"` but it didn't work.
It returns (`java.beans.IntrospectionException: Method not found: setRejectionPolicy`). Is there a way to overcome this problem?

**Comment by kwark** [comment]
I think this issue is related to https://github.com/slick/slick/issues/1274 and was closed prematurely.
The fact that people complain about this only happening when running with transactions was never explained and to easily dismissed.

In that other issue a quick succession of transactional DBIO actions causes the connection pool to be completely consumed and the slick executor threads are blocked during the connection timeout period.
If load keeps getting applied then the executor queue keeps growing.

So, in my case I first see a couple of SqlTimeoutExceptions followed by an onslaught of RejectedExecutionException.

**Comment by RichyHBM** [comment]
Did this ever get resolved? 

I am seeing the same issue when unit testing with Specs, using Play 2.4 and Play-Slick 1.1.1

FYI I am using a FakeApplication in my tests, which may be where the error comes from as this doesn't happen when running via play normally

```
def fakeApp: FakeApplication = FakeApplication(additionalConfiguration = Map(
  "slick.dbs.default.driver" -> "slick.driver.H2Driver$",
  "slick.dbs.default.db.driver" -> "org.h2.Driver",
  "slick.dbs.default.db.url" -> "jdbc:h2:mem:test;MODE=PostgreSQL;DATABASE_TO_UPPER=FALSE"
))  
```

**Comment by JavierCane** [comment]
I have this very same issue. In my case, I've implemented a batch migration script using standard streams just because of the back-pressure feature, so I was expecting to avoid this kind of errors.

Here you have the implementation in order to read from a huge table and insert the records in a destination database. Both of them running MySQL:

```scala
    val query = for {
      c <- SourceTableRow
      p <- SourceTable2Row if c.foreignKey === p.primaryKey
    } yield (c, p.otherField)

    def enableStream(statement: java.sql.Statement): Unit = {
      statement match {
        case s: com.mysql.jdbc.StatementImpl => s.enableStreamingResults()
        case _ =>
      }
    }

    val publisher = sourceDb.stream(query.result.withStatementParameters(statementInit = enableStream))

    val source = Source.fromPublisher(publisher)

    val saveResult = source.runForeach {
      case (sourceTable1Row, sourceTable2Field) =>

        val model = DestinationModel(
          // ...
        )

        val result = destinationDb.run(Persistence.model += model)
```

As I said, I was expecting to have the back-pressure feature by default. Am I wrong? how do I have to deal with this?

If I increase the allowed amount of task in queue, I'll be only moving the problem forward because it's a matter of time that the script reaches this limitation :-/

**Comment by szeiger** [comment]
@JavierCane I think execution based on consuming a Reactive Stream instead of calling `db.run` or `db.stream` directly would be useful. We could propagate back-pressure to the caller to prevent the queue from overflowing. The current design doesn't really take this into account. It was built for an interactive / web / web service app where requests come in randomly and you want them to fail quickly under load, with the queue size controlling the trade-off between availability and latency. Could you open a separate ticket for this?

**Comment by dsounded** [comment]
Have the same with my specs: just after `FakeRequest` I've got the error, `DELETE/INSERT` actions, but `SELECT` is OK

**Comment by dsounded** [comment]
@szeiger Any workaround to get rid of this in specs ? Because I am not sure if boosting thread pool is correct solution

**Comment by dsounded** [comment]
If you use `OneAppPerTes`t, replace it with `OneAppPerSuite`

**Comment by pedrorijo91** [comment]
@RichyHBM how did you overcome your issue? having the same problem on specs with fake application...

```
def fakeApplication = FakeApplication(additionalConfiguration =
    Map(
      "slick.dbs.default.driver" -> "slick.driver.H2Driver$",
      "slick.dbs.default.db.driver" -> "org.h2.Driver",
      "slick.dbs.default.db.url" -> "jdbc:h2:mem:play"
    ))
```

**Comment by pedrorijo91** [comment]
instead of running the fake application for each test I'm using `step(Play.start(fakeApplication))` at the beginning of the test suite and `step(Play.stop(fakeApplication))` at the end. 

seems to work (because it's always the same application/db connection I suppose) - in case someone needs a workaround

**Comment by cvogt** [comment]
@d6y do you think this could be solved by better documentation? Reopening this as a documentation issue for now.

**Comment by RichyHBM** [comment]
@pedrorijo91 That is pretty much how I solved it, just have the 1 play application running for all of your tests

**Comment by dsounded** [comment]
Yeah, `stop/start` workaround works for me as well, not a brilliant solution, but anyway

**Comment by pedrorijo91** [comment]
the problem is that the test database (H2) is dirty between each test on the test suite...but I guess I can clear (all) tables, or split into several test suite

**Comment by dsounded** [comment]
Yeah, I've done my small `DatabaseCleaner` for this stuff, which gets `List` of tables which should be truncated

**Comment by yurique** [comment]
With Slick 3.2 (hikaricp, default settings) we had this:
```
Task slick.basic.BasicBackend$DatabaseDef$$anon$2@dc100cf rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@2f716c9e[Running, pool size = 20, active threads = 0, queued tasks = 0, completed tasks = 74259]
```

and it was not recovering from it, had to restart app server.

Is there a cure? :)

**Comment by abdelrahmanmohamed** [comment]
+1

**Comment by acjay** [comment]
I'm also seeing this when running my tests, very predictably.

**Comment by sourcekick** [comment]
We are also running into this. Is it already supported to increase the `queueSize` (which has been mentioned at a potential cure here) and other related settings like number of Threads when using the convenient `DatabaseConfig` described at http://slick.lightbend.com/doc/3.2.1/database.html#databaseconfig ?

**Comment by sourcekick** [comment]
I think I found the answer is yes at http://slick.lightbend.com/doc/3.2.1/api/index.html#slick.jdbc.JdbcBackend$DatabaseFactoryDef `forConfig`.
